src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java

```diff
@@ -110,12 +118,62 @@ protected Object readResolve() throws ObjectStreamException {
      * It will also be false for something that has finished but is pending child node creation,
      * such as a completed fork branch which is waiting for the join node to be created.
      * <p>This can only go from true to false and is a shortcut for {@link FlowExecution#isCurrentHead}.
+     * @see #isActive
      */
     @Exported
     public final boolean isRunning() {
         return getExecution().isCurrentHead(this);
     }
 
+    private static final Map<FlowExecution, Set<BlockStartNode>> unclosedBlocks = new WeakHashMap<>();
+    @Restricted(DoNotUse.class)
+    @Extension public static class BlockListener implements GraphListener.Synchronous {
+        @Override public void onNewHead(FlowNode node) {
+            if (node instanceof BlockStartNode || node instanceof BlockEndNode) {
+                FlowExecution exec = node.getExecution();
+                synchronized (unclosedBlocks) {
+                    Set<BlockStartNode> blocks = unclosedBlocks.get(exec);
+                    if (blocks == null) {
+                        unclosedBlocks.put(exec, blocks = new HashSet<>());
+                    }
+                    if (node instanceof BlockStartNode) {
+                        blocks.add((BlockStartNode) node);
+                    } else {
+                        blocks.remove(((BlockEndNode) node).getStartNode());
+                    }
+                }
+            }
+        }
+    }
+    /**
+     * Checks whether a node is still part of the active part of the graph.
+     * Unlike {@link #isRunning}, this behaves intuitively for a {@link BlockStartNode}:
+     * it will be considered active until the {@link BlockEndNode} is added.
+     */
+    public final boolean isActive() {
+        if (this instanceof FlowEndNode) { // cf. JENKINS-26139
+            return false;
+        } else if (this instanceof BlockStartNode) { // cf. JENKINS-38223
+            synchronized (unclosedBlocks) {
+                Set<BlockStartNode> blocks = unclosedBlocks.get(exec);
+                if (blocks != null) {
+                    return blocks.contains((BlockStartNode) this);
```

Comment:
Note that while initially it would seem that this check fully covers the case where you have restated the master with an in-progress flow and lost the map, there's a potential to be inside a long running block (for example a parallel running automation)... and then begin another block, which populates the map... but now you're missing the block start for the parallel start.